docs/src/design.md

```diff
@@ -0,0 +1,175 @@
+They should only be called as the final step in the previous layer.
+
+If the default parameters are not dispatched per type, using `_` might be skipped.
+The following resolution might even be seen as a last step in layer I or the resolution here in layer II.
+
+```julia
+exp(M::AbstractManifold, p, X, t::Real) = exp(M, p, t * X)
+```
+
+When there is no dispatch for different types of the optional parameter (here `t`), the `_` might be skipped.
+One could hence see the last code line as a definition on Layer I that passes directly to Layer III, since there are not parameter to dispatch on.
+
+To close this section, let‘s look at an example. The high level (or level I) definition of the retraction is given by
+
+```julia
+retract(M::AbstractManifold, p, X, m::AbstractRetractionMethod=default_retraction_method(M)) = _retract(M,p,X,m)
+```
+
+This level now dispatches on different retraction types. It usually passes to specific functions implemented in Layer III,
+here for example
+
+```julia
+_retract(M::AbstractManifold, p, X, m::Exponentialretraction) = exp(M,p,X)
+_retract(M::AbstractManifold, p, X, m::PolarRetraction) = retract_polar(M,p,X)
+```
+
+or the [`PolarRetraction`](@ref) which dispatches to [`retract_polar`](@ref).
+
+For further details and dispatches, see [retractions and inverse retractions](@ref sec-retractions) for an overview.
+
+!!! note
+    The documentation should be attached to the high level functions, since this again fosters ease of use.
+    If yuo implement a polar retraction, you should write a function `polar_retract` but the doc string should be attached to `retract(::M, ::P, ::V, ::PolarRetraction)` for your types `::M, ::P, ::V` of the manifold, points and vectors, respectively.
+
+To summarize, with respect to the [dispatch on one argument at a time](https://docs.julialang.org/en/v1/manual/methods/#Dispatch-on-one-argument-at-a-time) paradigm, this layer dispatches the (optional) _parameters second_.
+
+### [Layer III: The base layer with focus on implementations](@id design-layer3)
+
+This lower level aims for the actual implementation of the function avoiding ambiguities.
+It should have as few as possible optional parameters and as concrete as possible types.
+This means
+
+* the function name should be similar to its high level parent (for example `retract` and `retract_polar`  above)
+* The manifold type in method signature should always be as narrow as possible.
+* the points/vectors should either be untyped (for the default representation of if there is only one implementation) or provide all types concretely (for second representations or when using [`AbstractManifoldPoint`](@ref) and [`TVector`](@ref TVector)).
+
+The first step that might happen on this level is, that an allocating function allocates memory and calls the mutating function. If faster, it might also implement the function at hand itself.
```

Comment:
```suggestion
The first step that often happens on this level is memory allocation and calling the mutating function. If faster, it might also implement the function at hand itself.
```
What is the second sentence supposed to mean?